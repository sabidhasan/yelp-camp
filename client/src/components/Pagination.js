import React from 'react'

const Pagination = (props) => {
  if (props.lastPage === 0) return <h2>No results found. Try applying a different filter</h2>;

  // Generate pages counter with current page already in there
  const allPages = [props.currentPage]
  for (let i = 1; i <= 10; i++) {
    // look current page +/- i
    const newNums = [props.currentPage + i, props.currentPage - i]
    // Add if allPages isnt too long and page number is not too small/big
    newNums.forEach(v => {if (v > 0 && v <= props.lastPage && allPages.length < 7) allPages.push(v)})
  }
  // Sort because it is in no order
  allPages.sort((a, b) => (a > b ? 1 : -1))

  return(
    <React.Fragment>
      <div className='search-page-selector'>
        <span className='page-count'>
          Page {props.currentPage} of {props.lastPage}
        </span>
        {allPages.length > 1 ?
          <div className='pages'>
            {props.currentPage > 1
              ? <a href='' className='prev-link' onClick={props.prevHandler}>« Previous</a>
              : null
            }
            {allPages.map(v=> (
              <a
                key={v}
                className={v===props.currentPage ? 'current' : undefined}
                onClick={(e) => props.goToPageHandler(e, v - 1)} href='#'>{v}
              </a>)
            )}
            {props.currentPage !== props.lastPage
              ? <a href='' className='next-link' onClick={props.nextHandler}>Next »</a>
              : null
            }
          </div>
        : null }
      </div>
  </React.Fragment>
  )
}

export default Pagination
