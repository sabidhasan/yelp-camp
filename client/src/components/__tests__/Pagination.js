import React from 'react'
import toJson from 'enzyme-to-json'
import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import Pagination from '../Pagination'

configure({ adapter: new Adapter() });

describe('<Pagination />', () => {
  let component;
  let props;
  const goToPageHandler = jest.fn();

  const propsGenerator = (currentPage, lastPage) => {
    return {
      currentPage,
      lastPage,
      goToPageHandler
    }
  }

  beforeEach(() => {
    props = {goToPageHandler}
    component = shallow(<Pagination { ...props } />)
  })

  it('Component should be defined', () => {
    expect(component).toBeDefined();
  });

  it('When lastPage is 0, nothing gets rendered', () => {
    props = propsGenerator(5, 0);
    component = shallow(<Pagination { ...props } />)
    expect(component.html()).toBeNull();
  });

  it('When currentPage is undefined, nothing gets rendered', () => {
    props = propsGenerator(undefined, 10);
    component = shallow(<Pagination { ...props } />)
    expect(component.html()).toBeNull();
  });

  it('Pagination count is correctly displayed', () => {
    props = propsGenerator(3, 10);
    component = shallow(<Pagination { ...props } />)
    expect(component.find('.Pagination__count').text()).toBe('Page 3 of 10')
  });

  it('If currentPage is > 1, it shows previous link', () => {
    props = propsGenerator(2, 10);
    component = shallow(<Pagination { ...props } />)
    expect(component.find('.Pagination__prev-link')).toHaveLength(1)
  });

  it('If currentPage is not > 1, it does not show previous link', () => {
    props = propsGenerator(1, 10);
    component = shallow(<Pagination { ...props } />)
    expect(component.find('.Pagination__prev-link')).toHaveLength(0)
  });

  it('If currentPage matches lastPage, it doesn\'t show next link', () => {
    props = propsGenerator(10, 10);
    component = shallow(<Pagination { ...props } />)
    expect(component.find('.Pagination__next-link')).toHaveLength(0)
  });

  it('If currentPage does not match lastPage, it shows next link', () => {
    props = propsGenerator(5, 10);
    component = shallow(<Pagination { ...props } />)
    expect(component.find('.Pagination__next-link')).toHaveLength(1)
  });

  it('Number of Pagination__link links is correct - 9 expected at maximum', () => {
    props = propsGenerator(2, 8);
    component = shallow(<Pagination { ...props } />)
    expect(component.find('.Pagination__link')).toHaveLength(9)
  });

  it('If only one page, then no pagination shown', () => {
    props = propsGenerator(1, 1);
    component = shallow(<Pagination { ...props } />)
    expect(component.find('.Pagination__link')).toHaveLength(0)
  });


  it('Function called when button clicked', () => {
    props = propsGenerator(1, 2);
    component = shallow(<Pagination { ...props } />)
    component.find('.Pagination__link').first().simulate('click');
    expect(goToPageHandler).toHaveBeenCalledTimes(1)
  })

  it('Current page link is correctly highlighted', () => {
    props = propsGenerator(5, 8);
    component = shallow(<Pagination { ...props } />)
    expect(component.find('.Pagination__link--current').text()).toBe('5')
  })

  it('Pagination component snapshot matches what is expected', () => {
    props = propsGenerator(1, 5);
    component = shallow(<Pagination { ...props } />)
    expect(toJson(component)).toMatchSnapshot();
  })
})


// currentPage matches Pagination__link--current's text()
