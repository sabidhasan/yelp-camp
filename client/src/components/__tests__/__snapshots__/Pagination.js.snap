// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> Pagination component snapshot matches what is expected 1`] = `
<div
  className="Pagination flex-center"
>
  <span
    className="Pagination__count"
  >
    Page 
    1
     of 
    5
  </span>
  <div
    className="Pagination__pages"
  >
    <button
      className="Pagination__link bold Pagination__link--current"
      key="1"
      onClick={[Function]}
    >
      1
    </button>
    <button
      className="Pagination__link bold "
      key="2"
      onClick={[Function]}
    >
      2
    </button>
    <button
      className="Pagination__link bold "
      key="3"
      onClick={[Function]}
    >
      3
    </button>
    <button
      className="Pagination__link bold "
      key="4"
      onClick={[Function]}
    >
      4
    </button>
    <button
      className="Pagination__link bold "
      key="5"
      onClick={[Function]}
    >
      5
    </button>
    <button
      className="Pagination__link Pagination__next-link bold"
      onClick={[Function]}
    >
      Next Â»
    </button>
  </div>
</div>
`;
